shadowsocks/client_test.go

```diff
@@ -0,0 +1,244 @@
+			}
+			defer clientConn.Close()
+			go func() {
+				ssr := NewShadowsocksReader(clientConn, cipher)
+				ssw := NewShadowsocksWriter(clientConn, cipher)
+				ssClientConn := onet.WrapConn(clientConn, ssr, ssw)
+
+				tgtAddr, err := socks.ReadAddr(ssClientConn)
+				if err != nil {
+					t.Fatalf("Failed to read target address: %v", err)
+				}
+				if tgtAddr.String() != expectedTgtAddr {
+					t.Fatalf("Expected target address '%v'. Got '%v'", expectedTgtAddr, tgtAddr)
+				}
+				io.Copy(ssw, ssr)
+			}()
+		}
+	}()
+	return listener.Addr()
+}
+
+func startShadowsocksUDPEchoServer(expectedTgtAddr string, t testing.TB) net.Addr {
+	conn, err := net.ListenUDP("udp", &net.UDPAddr{IP: net.ParseIP("127.0.0.1"), Port: 0})
+	if err != nil {
+		t.Fatalf("Proxy ListenUDP failed: %v", err)
+	}
+	t.Logf("Starting SS UDP echo proxy at %v\n", conn.LocalAddr())
+	cipherBuf := make([]byte, udpBufSize)
+	clientBuf := make([]byte, udpBufSize)
+	cipher, err := newAeadCipher(testCipher, testPassword)
+	if err != nil {
+		t.Fatalf("Failed to create cipher: %v", err)
+	}
+	go func() {
+		defer conn.Close()
+		for {
+			n, clientAddr, err := conn.ReadFromUDP(cipherBuf)
+			if err != nil {
+				t.Fatalf("Failed to read from UDP conn: %v", err)
+			}
+			buf, err := shadowaead.Unpack(clientBuf, cipherBuf[:n], cipher)
+			if err != nil {
+				t.Fatalf("Failed to decrypt: %v", err)
+			}
+			tgtAddr := socks.SplitAddr(buf)
+			if tgtAddr == nil {
+				t.Fatalf("Failed to read target address: %v", err)
+			}
+			if tgtAddr.String() != expectedTgtAddr {
+				t.Fatalf("Expected target address '%v'. Got '%v'", expectedTgtAddr, tgtAddr)
+			}
+			// Echo both the payload and SOCKS address.
+			buf, err = shadowaead.Pack(cipherBuf, buf, cipher)
+			if err != nil {
+				t.Fatalf("Failed to encrypt: %v", err)
+			}
+			conn.WriteTo(buf, clientAddr)
+			if err != nil {
+				t.Fatalf("Failed to write: %v", err)
+			}
+		}
+	}()
+	return conn.LocalAddr()
+}
+
+// io.ReadWriter adapter for net.PacketConn. Used to share code between UDP and TCP tests.
+type packetConnReadWriter struct {
+	net.PacketConn
+	io.ReadWriter
+	targetAddr net.Addr
+}
+
+func (pc *packetConnReadWriter) Read(b []byte) (n int, err error) {
+	n, _, err = pc.PacketConn.ReadFrom(b)
+	return
+}
+
+func (pc *packetConnReadWriter) Write(b []byte) (int, error) {
+	return pc.PacketConn.WriteTo(b, pc.targetAddr)
+}
+
+func expectEchoPayload(conn io.ReadWriter, payload, buf []byte, t testing.TB) {
+	_, err := conn.Write(payload)
+	if err != nil {
+		t.Fatalf("Failed to write payload: %v", err)
+	}
+	n, err := conn.Read(buf)
+	if err != nil {
+		t.Fatalf("Failed to read payload: %v", err)
+	}
+	if !bytes.Equal(payload, buf[:n]) {
+		t.Fatalf("Expected output '%v'. Got '%v'", payload, buf[:n])
+	}
+}
+
+// splitHostPortNumber parses the host and port from `address`, which has the form `host:port`,
+// validating that the port is a number.
+func splitHostPortNumber(address string) (host string, port int, err error) {
```

Comment:
A similar function is exported from `dialer.go`.  Is it necessary to duplicate it here?

Reply:
N/A, see shadowsocks/client.go